A46-xds-nack-semantics-improvement.md

```diff
@@ -0,0 +1,119 @@
+that an individual resource that sets a supported field to an unsupported
+value will be considered invalid; that behavior is intentional, and there
+is nothing we can do to prevent the need for clients to be upgraded to
+use a configuration resource that requires features that they do not yet
+support.  However, this document *is* intending to solve the problem of
+that one invalid resource causing other *valid* resources to be ignored.
+
+This document proposes a behavior change to address this problem.
+
+### Related Proposals: 
+* [gRFC A27: xDS-Based Global Load Balancing](A27-xds-global-load-balancing.md)
+* [gRFC A42: xDS Ring Hash LB Policy](A42-xds-ring-hash-lb-policy.md)
+
+## Proposal
+
+We will change gRPC's behavior such that when a response is NACKed, gRPC
+will still use all valid resources from the response; it will ignore
+only the invalid resources.
+
+Note that the xDS wire protocol behavior is not changing at all; the
+protocol currently still requires NACKs to be done on a per-response
+basis instead of a per-resource basis (although the latter is something
+that is expected to be added to the protocol in the future).  The only
+change is that the client will actually use the valid resources in the
+response.
+
+### Temporary environment variable protection
+
+N/A
+
+## Rationale
+
+Note that gRPC's original behavior was intended to make sure that the
+control plane had a clear picture of what configuration the client is
+using.  However, it turns out that Envoy currently has cases where it
+will apply some valid resources from a response that it NACKs, which
+means that the control plane *already* did not have that kind of clear
+picture.  To ensure that the control plane will have that kind of clear
+picture, there will be subsequent work to extend the xDS protocol to
+allow per-resource NACKing instead of per-response NACKing, although
+that will be a broader project.
+
+We considered the following alternatives:
+
+- We could have just waited for the xDS protocol changes that will allow
+  per-resource NACKing instead of per-response NACKing.  However, that
+  is likely to take more work to design and implement in both clients
+  and control planes, and we wanted to do something quickly to alleviate
+  the problem described above.
+
+- For the case of unknown LB policies specifically, we could fall back
+  to `ROUND_ROBIN` instead of considering the resource invalid.
+  However, this is not semantically correct behavior; it could cause
+  unexpected load on servers and a large number of unnecessary connections
+  to backends.  Also, while it would address the specific case that
+  triggered this design, it would not address the general case of seeing
+  an unsupported value in a supported field.
+
+## Implementation
+
+[A description of the steps in the implementation, who will do them, and when.  If a particular language is going to get the implementation first, this section should list the proposed order.]
```

Comment:
TBD? Remove?

Reply:
I'll update this section once I have an implementation PR for C-core out for review.